jujugui/static/gui/src/app/app.js

```diff
@@ -669,6 +662,112 @@ YUI.add('juju-gui', function(Y) {
     },
 
     /**
+      Creates a sandbox connection backend for use with the GoJujuAPI
+
+      @method createSandboxConnection
+      @param {Object} cfg The configuration object to use and modify with the
+        sandbox connection instance.
+      @return {Object} cfg A modified version of the passed in cfg which
+        constains the sandbox connection.
+    */
+    createSandboxConnection: function(cfg) {
+      cfg.socket_url = this.get('sandboxSocketURL');
+      let environments = juju.environments;
+      let sandboxModule = environments.sandbox;
+      const fakebackend = new environments.FakeBackend({
+        charmstore: this.get('charmstore')
+      });
+      if (cfg.user && cfg.password) {
+        let credentials = fakebackend.get('authorizedUsers');
+        credentials['user-' + cfg.user] = cfg.password;
+        fakebackend.set('authorizedUsers', credentials);
+      }
+      cfg.conn = new sandboxModule.ClientConnection({
+        juju: new sandboxModule.GoJujuAPI({
+          state: fakebackend,
+          socket_url: cfg.socket_url
+        })
+      });
+      // Instantiate a fake Web handler, which simulates the
+      // request/response communication between the GUI and the juju-core
+      // HTTPS API.
+      cfg.webHandler =
+        new environments.web.WebSandbox({state: fakebackend});
+      return cfg;
+    },
+
+    /**
+      Creates the second instance of the websocket for communication with
+      the Juju controller if it's necessary. A username and password must be
+      supplied if you're connecting to a standalone Juju controller and not to
+      one which requires macaroon authentication.
+
+      @method setUpControllerAPI
+      @param {Object} controllerAPI Instance of the GoEnvironment class.
+      @param {String} user The username if not using macaroons.
+      @param {String} password The password if not using macaroons.
+      @param {Array} macaroons A list of macaroons that the user has saved.
+      @return {environments.GoEnvironment} An instance of the environment class
+        with the necessary events attached and values set. Or undefined if
+        we're in a legacy juju model and no controllerAPI instance was supplied.
+    */
+    setUpControllerAPI: function(controllerAPI, user, password, macaroons) {
+      // If no instance was supplied to us then we're in a legacy Juju
+      // and this connection is not necessary.
+      if (!controllerAPI) { return; }
+
+      controllerAPI.setAttrs({ user, password });
+      controllerAPI.setCredentials({ user, password, macaroons });
+
+      controllerAPI.after('login', e => {
+        this.controllerAPI.listModelsWithInfo((err, response) => {
+          if (err) {
+            console.error('Unable to list models', err);
```

Comment:
Let's use a lower case convention for errors: `console.error('unable to list models:', err);`?
[shrug]